lib/improver/spotdata/neighbour_finding.py

```diff
@@ -0,0 +1,445 @@
+                         is needed.
+                         Must contain {'orography': <cube of orography>}.
+                         Needs {'land': <cube of land mask>} if using land
+                         constraint.
+
+        Returns:
+        --------
+        neighbours: Numpy array of grid i,j coordinates that are nearest to
+                    each site coordinate given. Includes height difference
+                    between site and returned grid point.
+
+        '''
+        # Use the default nearest neighbour list as a starting point, and
+        # if for some reason it is missing, recreate the list using the fast
+        # method.
+        if default_neighbours is None:
+            neighbour_list = self.fast_nearest_neighbour(cube, sites,
+                                                         ancillary_data)
+        else:
+            neighbour_list = default_neighbours
+
+        orography = data_from_ancillary(ancillary_data, 'orography')
+        if land_constraint:
+            land = data_from_ancillary(ancillary_data, 'land')
+
+        for i_site, site in enumerate(sites.itervalues()):
+            altitude = site['altitude']
+
+            i, j = neighbour_list['i'][i_site], neighbour_list['j'][i_site]
+            edgecase = neighbour_list['edge'][i_site]
+
+            node_list = nearest_n_neighbours(i, j, 9)
+            if edgecase:
+                node_list = node_edge_test(node_list, cube)
+
+            if land_constraint:
+                # Check that we are considering a land point and that at least
+                # one neighbouring point is also land. If not no modification
+                # is made to the nearest neighbour coordinates.
+
+                exclude_self = nearest_n_neighbours(i, j, 9, exclude_self=True)
+                if edgecase:
+                    exclude_self = node_edge_test(exclude_self, cube)
+                if not land[i, j] or not any(land[exclude_self]):
+                    continue
+
+                node_list = ConditionalListExtract('not_equal_to').process(
+                    land, node_list, 0)
+
+            dz_nearest = abs(altitude - orography[i, j])
+            dzs = altitude - orography[node_list]
+
+            dzs, dz_nearest, dz_subset = apply_bias(
+                relative_z, dzs, dz_nearest, altitude, orography, i, j)
+
+            ij_min = index_of_minimum_difference(dzs, subset_list=dz_subset)
+            i_min, j_min = list_entry_from_index(node_list, ij_min)
+            dz_min = abs(altitude - orography[i_min, j_min])
+
+            if dz_min < dz_nearest:
+                neighbour_list[i_site] = i_min, j_min, dzs[ij_min], edgecase
+
+        return neighbour_list
+
+# Wrapper routines to use the dz minimisation routine with various options.
+# These can be called as methods and set in the diagnostic configs.
+# It may be better to simply use the keyword options at a higher level,
```

Comment:
I've removed the various predefined options and now included the setting of options in my configuration dictionary.